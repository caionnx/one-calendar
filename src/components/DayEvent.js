import React, {Component} from 'react';
import Modal from 'react-modal';

import '../styles/Modal.scss';
import '../styles/Button.scss';

class DayEvent extends Component {
  constructor () {
    super();

    this.state = {
      selectedCategory: false,
      defaultCategories: {
        holiday: 'Holiday',
        birthday: 'Birthday',
        busy: 'Busy',
        anniversary: 'Anniversary'
      }
    }
  }

  onChangeCategory (ev) {
    const value = ev.target.value
    this.setState({
      selectedCategory: { id: value, name: this.state.defaultCategories[value] }
    })
  }

  resetSelected () {
    this.setState({
      selectedCategory: false
    })
  }

  render () {
    const {
      modalIsOpen,
      closeModal,
      fullDate,
      event,
      onRemoveEvent,
      onAddEvent,
    } = this.props;

    return (
        <Modal
          className='Modal'
          appElement={document.querySelector('#root')}
          isOpen={modalIsOpen}
          onRequestClose={closeModal}
          contentLabel={fullDate}>

          <div className='Modal__Flex'>
            <h4 className='Modal__Title'>{fullDate}</h4>
            <button
              aria-label='Close modal'
              className='Button Button--Unstyled Button--Large'
              onClick={closeModal}>X</button>
          </div>
          {
            event
            ? <div className='Modal__Content Modal__Flex'>
                {event.name}
                <button className='Button Button--Delete' onClick={() => {
                  this.resetSelected();
                  onRemoveEvent(event.id);
                  }}>
                  Delete event
                </button>
              </div>
            : <form className='Modal__Content Modal__Flex' onSubmit={ev => onAddEvent(ev, this.state.selectedCategory)}>
                <select onChange={ev => this.onChangeCategory(ev)}>
                  <option key='default'>Select one option</option>
                  {Object.keys(this.state.defaultCategories).map(ct => 
                    <option key={ct} value={ct}>{this.state.defaultCategories[ct]}</option>
                  )}
                </select>
                <button className='Button'>Add event</button>
              </form>
          }
        </Modal>
    )
  }
}

export default DayEvent;