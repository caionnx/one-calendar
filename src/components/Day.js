import React, { Component } from 'react';
import dateFormat from 'date-fns/format';
import isToday from 'date-fns/is_today';
import { connect } from 'react-redux'

import { addEvent, removeEvent } from '../actions'
import DayEvent from './DayEvent';

const getUID = ({ numeric, month, year } = {}) =>
  `${numeric}${month}${year}`;

export class Day extends Component {
  constructor () {
    super();

    this.state = {
      event: false,
      eventModal: false,
    }

    this.openEventModal = this.openEventModal.bind(this);
    this.closeEventModal = this.closeEventModal.bind(this);
  }

  componentDidMount () {
    const { year, month, numeric } = this.props;
    const dirtyFormat = `${year}-${month + 1}-${numeric}`;

    this.setState({ fullDate: dateFormat(dirtyFormat, 'D MMM YYYY') });
  }

  openEventModal () {
    this.setState({ eventModal: true });
  }

  closeEventModal (ev) {
    ev.preventDefault();
    this.setState({ eventModal: false });
  }

  // Define class name and modifiers e.g. Weekend, Current, Holiday etc.
  getClassName (isVisible) {
    const { event } = this.props;
    const current = this.isToday() ? ' Day--current' : '';
    const highlighted = event ? ` Day--${event.id}` : '';
    const weekend = this.isWeekend() ? ' Day--weekend' : '';
    const visibility = !isVisible ? ' Day--hidden' : '';

    return `Day${highlighted}${current}${weekend}${visibility}`;
  }

  isToday () {
    const { year, month, numeric } = this.props;
    const dirtyFormat = `${year}-${month + 1}-${numeric}`;

    return isToday(dirtyFormat);
  }

  isWeekend () {
    const { dayIndex } = this.props;

    return dayIndex === 0 || dayIndex === 6;
  }

  onRemoveEvent () {
    const dateUID = getUID(this.props);

    this.props.dispatch(removeEvent(dateUID));
  }

  onAddEvent (ev, category) {
    ev.preventDefault();
    const { id, name } = category;
    const dateUID = getUID(this.props);

    if (!id || !name) {
      return;
    }

    this.props.dispatch(addEvent(id, name, dateUID));
  }

  render () {
    const { numeric, event } = this.props;
    const { eventModal, fullDate } = this.state;
    
    return (
      <td>
        {numeric >= 1 ?
          [
            <div
              key={numeric}
              className={this.getClassName(numeric >= 1)}
              onClick={this.openEventModal}>
              { numeric }
            </div>,
            eventModal &&
            <DayEvent
              key={fullDate}
              event={event}
              fullDate={fullDate}
              modalIsOpen={eventModal}
              onRemoveEvent={id => this.onRemoveEvent(id)}
              onAddEvent={(ev, category) => this.onAddEvent(ev, category)}
              closeModal={ev => this.closeEventModal(ev)} />
          ]
          : ''
        }
      </td>
    );
  }
}

const mapStateToProps = ({ events }, props) => ({
  event: events.find(ev => ev.date === getUID(props))
});

export default connect(mapStateToProps)(Day);