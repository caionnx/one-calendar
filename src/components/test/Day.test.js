import React from 'react';
import { shallow } from 'enzyme';

import DayEvent from '../DayEvent';
import { Day } from '../Day';

let wrapper;
let propDispatch;
const props = {
  numeric: 15,
  month: 11,
  year: 2018
}

it('render DayEvent on click', () => {
  wrapper = shallow(<Day {...props} />);
  wrapper.find('div').at(0).simulate('click');

  expect(wrapper.find(DayEvent)).toHaveLength(1);
});

it('render day correctly (with DayEvent)', () => {
  wrapper = shallow(<Day {...props} />);
  wrapper.setState({ eventModal: true });

  expect(wrapper.find(DayEvent)).toHaveLength(1);
  expect(wrapper).toMatchSnapshot();
});

it('execute dispatch from props on "add event"', () => {
  propDispatch = jest.fn();

  wrapper = shallow(<Day dispatch={propDispatch} {...props} />);
  wrapper.setState({ eventModal: true });
  wrapper.find(DayEvent).prop('onAddEvent')(
    { preventDefault: jest.fn() },
    { id: 'test', name: 'Test' }
  );

  expect(propDispatch).toHaveBeenCalled()
});

it('execute dispatch from props on "remove event"', () => {
  propDispatch = jest.fn();
  wrapper = shallow(<Day dispatch={propDispatch} {...props} />);
  wrapper.setState({ eventModal: true });
  wrapper.find(DayEvent).prop('onRemoveEvent')();

  expect(propDispatch).toHaveBeenCalled()
});

it('execute closeModal and change state', () => {
  wrapper = shallow(<Day {...props} />);
  wrapper.setState({ eventModal: true });
  wrapper.find(DayEvent).prop('closeModal')({ preventDefault: jest.fn() });

  expect(wrapper.state('eventModal')).toBeFalsy();
});