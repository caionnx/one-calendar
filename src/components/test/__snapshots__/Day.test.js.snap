// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render day correctly (with DayEvent) 1`] = `
<td>
  <div
    className="Day"
    key="15"
    onClick={[Function]}
  >
    15
  </div>
  <DayEvent
    closeModal={[Function]}
    fullDate="15 Dec 2018"
    key="15 Dec 2018"
    modalIsOpen={true}
    onAddEvent={[Function]}
    onRemoveEvent={[Function]}
  />
</td>
`;
